identity/keystore_filesystem.go

```diff
@@ -18,17 +18,189 @@
+// NewKeystoreFilesystem create new keystore, which keeps keys in filesystem.
+func NewKeystoreFilesystem(directory string, lightweight bool) *Keystore {
+	var ks *ethKs.KeyStore
 	if lightweight {
 		log.Debug().Msg("Using lightweight keystore")
-		return keystore.NewKeyStore(directory, keystore.LightScryptN, keystore.LightScryptP)
+		ks = ethKs.NewKeyStore(directory, ethKs.LightScryptN, ethKs.LightScryptP)
+	} else {
+		log.Debug().Msg("using heavyweight keystore")
+		ks = ethKs.NewKeyStore(directory, ethKs.StandardScryptN, ethKs.StandardScryptP)
+	}
+
+	return &Keystore{
+		ethKeystore: ks,
+		derivedKeys: make(map[common.Address]*memguard.Enclave),
+	}
+}
+
+// Keystore handles everything that's related to eth accounts.
+type Keystore struct {
+	ethKeystore *ethKs.KeyStore
+
+	derivedKeys    map[common.Address]*memguard.Enclave
+	derivedKeyLock sync.Mutex
+}
+
+// Accounts returns all accounts.
+func (ks *Keystore) Accounts() []accounts.Account {
+	return ks.ethKeystore.Accounts()
+}
+
+// NewAccount creates a new account.
+func (ks *Keystore) NewAccount(passphrase string) (accounts.Account, error) {
+	return ks.ethKeystore.NewAccount(passphrase)
+}
+
+// Find finds an account.
+func (ks *Keystore) Find(a accounts.Account) (accounts.Account, error) {
+	return ks.ethKeystore.Find(a)
+}
+
+// Unlock unlocks an account.
+func (ks *Keystore) Unlock(a accounts.Account, passphrase string) error {
+	err := ks.ethKeystore.Unlock(a, passphrase)
+	if err != nil {
+		return err
+	}
+
+	derived, err := ks.deriveKey(a, passphrase)
+	if err != nil {
+		return err
+	}
+
+	ks.rememberDerivedKey(a.Address, derived)
+	return nil
+}
+
+// Lock locks an account.
+func (ks *Keystore) Lock(addr common.Address) error {
+	defer ks.forgetDerivedKey(addr)
+	return ks.ethKeystore.Lock(addr)
+}
+
+func (ks *Keystore) rememberDerivedKey(a common.Address, key []byte) {
+	ks.derivedKeyLock.Lock()
+	defer ks.derivedKeyLock.Unlock()
+	enclave := memguard.NewEnclave(key)
+	ks.derivedKeys[a] = enclave
+}
+
+func (ks *Keystore) forgetDerivedKey(a common.Address) {
+	ks.derivedKeyLock.Lock()
+	defer ks.derivedKeyLock.Unlock()
+	delete(ks.derivedKeys, a)
+}
+
+func (ks *Keystore) getDerivedKey(a common.Address) ([]byte, error) {
+	ks.derivedKeyLock.Lock()
+	defer ks.derivedKeyLock.Unlock()
+
+	if _, ok := ks.derivedKeys[a]; !ok {
+		return nil, errors.New("no key found")
+	}
+
+	enclave := ks.derivedKeys[a]
+	buffer, err := enclave.Open()
+	if err != nil {
+		return nil, err
 	}
+	defer buffer.Destroy()
+
+	bytes := buffer.Bytes()
+
+	copied := make([]byte, len(bytes))
+	copy(copied, bytes)
+	return copied, nil
+}
+
+func (ks *Keystore) deriveKey(a accounts.Account, passphrase string) ([]byte, error) {
+	kjson, err := ks.ethKeystore.Export(a, passphrase, passphrase)
+	if err != nil {
+		return nil, err
+	}
+
+	k, err := ethKs.DecryptKey(kjson, passphrase)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
I suspect this private key might get left in memory decrypted. 
```suggestion
defer memguard.WipeBytes(k.PrivateKey.D.Bytes())
```

Reply:
It won't. Function vars won't escape to heap, meaning they will stay on the stack, which in turn means they will not get garbage collected since the stack is destroyed once the function returns. I can add the wipe if you prefer though.

Reply:
Destruction of a stack does not mean that physical memory will seize to contain this key, it will simply destroy references to it. Thats why wipe should be needed imho.